File: training-data/javascript/template_method_pattern.js
class DataExporter {
  export(records) {
    const header = this.formatHeader();
    const rows = records.map((r) => this.formatRow(r));
    return this.combine(header, rows);
  }
  formatHeader() {
    throw new Error('not implemented');
  }
  formatRow() {
    throw new Error('not implemented');
  }
  combine(header, rows) {
    return [header, ...rows].join('\n');
  }
}

class CsvExporter extends DataExporter {
  formatHeader() {
    return 'name,age';
  }
  formatRow(record) {
    return `${record.name},${record.age}`;
  }
}

class MarkdownExporter extends DataExporter {
  formatHeader() {
    return '| name | age |\n| --- | --- |';
  }
  formatRow(record) {
    return `| ${record.name} | ${record.age} |`;
  }
}

const records = [
  { name: 'Alice', age: 30 },
  { name: 'Bob', age: 25 },
];

console.log(new CsvExporter().export(records));
console.log(new MarkdownExporter().export(records));
module.exports = { DataExporter, CsvExporter, MarkdownExporter };

File: training-data/python/template_method_pattern.py
from abc import ABC, abstractmethod


class DataExporter(ABC):
    """Template method: fixed skeleton, customizable steps."""

    def export(self, records):
        header = self.format_header()
        rows = [self.format_row(r) for r in records]
        return "\n".join([header, *rows])

    @abstractmethod
    def format_header(self):
        ...

    @abstractmethod
    def format_row(self, record):
        ...


class CsvExporter(DataExporter):
    def format_header(self):
        return "name,age"

    def format_row(self, record):
        return f"{record['name']},{record['age']}"


class MarkdownTableExporter(DataExporter):
    def format_header(self):
        return "| name | age |\n|------|-----|"

    def format_row(self, record):
        return f"| {record['name']} | {record['age']} |"


if __name__ == "__main__":
    people = [{"name": "Ada", "age": 36}, {"name": "Alan", "age": 41}]
    print(CsvExporter().export(people))
    print()
    print(MarkdownTableExporter().export(people))
